main/src/ca/uwaterloo/flix/language/phase/ConstraintGeneration.scala

```diff
@@ -0,0 +1,1071 @@
+/*
+ * Copyright 2015-2023 Magnus Madsen, Matthew Lutze
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *   http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+package ca.uwaterloo.flix.language.phase
+
+import ca.uwaterloo.flix.api.Flix
+import ca.uwaterloo.flix.language.ast.KindedAst.Expr
+import ca.uwaterloo.flix.language.ast.Type.getFlixType
+import ca.uwaterloo.flix.language.ast.{Ast, Kind, KindedAst, Level, Name, RigidityEnv, Scheme, SemanticOp, SourceLocation, Symbol, Type, TypeConstructor}
+import ca.uwaterloo.flix.language.phase.constraintgeneration.{RestrictableChooseConstraintGeneration, SchemaConstraintGeneration, TypeContext, TypingConstraint}
+import ca.uwaterloo.flix.util.{InternalCompilerException, ParOps}
+
+object ConstraintGeneration {
+
+  /**
+    * Generates constraints for the given expression.
+    */
+  def visitExp(exp0: KindedAst.Expr)(implicit c: TypeContext, root: KindedAst.Root, flix: Flix): (Type, Type) = {
+    // Make the context's level available
+    // This is a def rather than a val because c is mutable.
+    implicit def level: Level = c.getLevel
+
+    exp0 match {
+      case Expr.Var(sym, loc) => (sym.tvar, Type.Pure)
+      case Expr.Def(sym, tvar, loc) =>
+        val defn = root.defs(sym)
+        val (tconstrs, defTpe) = Scheme.instantiate(defn.spec.sc, loc.asSynthetic)
+        c.unifyTypeM(tvar, defTpe, loc)
+        c.addClassConstraintsM(tconstrs, loc)
+        val resTpe = defTpe
+        val resEff = Type.Pure
+        (resTpe, resEff)
+
+      case Expr.Sig(sym, tvar, loc) =>
+        val sig = root.classes(sym.clazz).sigs(sym)
+        val (tconstrs, sigTpe) = Scheme.instantiate(sig.spec.sc, loc.asSynthetic)
+        c.unifyTypeM(tvar, sigTpe, loc)
+        c.addClassConstraintsM(tconstrs, loc)
+        val resTpe = sigTpe
+        val resEff = Type.Pure
+        (resTpe, resEff)
+
+      case Expr.Hole(sym, tpe, loc) =>
+        (tpe, Type.Pure)
+
+      case Expr.HoleWithExp(exp, tvar, evar, loc) =>
+        val (tpe, eff) = visitExp(exp)
+        // effect type is AT LEAST the inner expression's effect
+        val atLeastEff = Type.mkUnion(eff, Type.freshVar(Kind.Eff, loc.asSynthetic), loc.asSynthetic)
+        c.unifyTypeM(atLeastEff, evar, loc)
+        // result type is whatever is needed for the hole
```

Comment:
This is for hole-with-expression, which is different, I guess?

Reply:
Thanks for the info. I suggest we make this just like hole. Then, once we have sets and subeffecting, we get the right solution.

Reply:
I think the current code is the correct type rule. if you think of `???(???, someExpression: t \ ef, ???)` (where `???` has unknown effect) then it will have any type and effect `ef + alpha`

Reply:
listed